src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -106,61 +81,180 @@ impl fmt::Display for Mode {
+            if mir.local_decls[arg].ty.needs_drop(tcx, param_env) {
+                local_needs_drop.add(&arg);
+            }
+        }
+
+        DropChecker {
+            span: mir.span,
+            mir,
+            tcx,
+            param_env,
+            local_needs_drop,
+            needs_drop: false
+        }
+    }
+
+    /// Add the given type's qualification to `self.needs_drop`.
+    fn add_type(&mut self, ty: Ty<'tcx>) {
+        self.needs_drop = ty.needs_drop(self.tcx, self.param_env);
+    }
+
+    /// Within the provided closure, `self.needs_drop` will start
+    /// out false, and its value after the closure returns will
+    /// be combined with the value before the call to nest.
+    fn nest<F: FnOnce(&mut Self)>(&mut self, f: F) {
+        let original = self.needs_drop;
+        self.needs_drop = false;
+        f(self);
+        self.needs_drop |= original;
+    }
+
+    /// Assign the current qualification to the given destination.
+    fn assign(&mut self, dest: &Place<'tcx>, _: Location) {
+        if let Place::Local(index) = *dest {
+            debug!("store to {:?} {:?}", self.mir.local_kind(index), index);
+            self.local_needs_drop.set_member(&index, self.needs_drop);
+        }
+    }
+}
+
+impl<'a, 'tcx> ConstChecker<'a, 'tcx, 'tcx> {
     fn new(tcx: TyCtxt<'a, 'tcx, 'tcx>,
            def_id: DefId,
            mir: &'a Mir<'tcx>,
            mode: Mode)
-           -> Qualifier<'a, 'tcx, 'tcx> {
+           -> ConstChecker<'a, 'tcx, 'tcx> {
         let mut rpo = traversal::reverse_postorder(mir);
         let temps = promote_consts::collect_temps(mir, &mut rpo);
         rpo.reset();
 
         let param_env = tcx.param_env(def_id);
 
-        let mut local_qualif = IndexVec::from_elem(None, &mir.local_decls);
+        let mut local_not_const = IdxSetBuf::new_filled(mir.local_decls.len());
         for arg in mir.args_iter() {
-            let mut qualif = Qualif::NEEDS_DROP;
-            qualif.restrict(mir.local_decls[arg].ty, tcx, param_env);
-            local_qualif[arg] = Some(qualif);
+            local_not_const.remove(&arg);
         }
 
-        Qualifier {
+        let mut_interior_checker = MutInteriorChecker::new(tcx, def_id, mir);
+        let drop_checker = DropChecker::new(tcx, def_id, mir);
+
+        ConstChecker {
             mode,
             span: mir.span,
             def_id,
             mir,
             rpo,
             tcx,
             param_env,
-            local_qualif,
-            qualif: Qualif::empty(),
+            local_not_const,
+            not_const: false,
+            mut_interior_checker,
+            drop_checker,
             const_fn_arg_vars: BitVector::new(mir.local_decls.len()),
             temp_promotion_state: temps,
-            promotion_candidates: vec![]
+            promotion_candidates: Vec::new()
         }
     }
 
+    /// Returns whether all qualification flags are unset.
+    fn is_qualif_empty(&self) -> bool {
```

Comment:
This is a hazard because it checks the other two flags without capturing the "instantaneous state".